src/Castle.Core/DynamicProxy/CustomAttributeInfo.cs

```diff
@@ -0,0 +1,238 @@
+				}
+
+				newExpression = memberInitExpression.NewExpression;
+
+				foreach (var binding in memberInitExpression.Bindings)
+				{
+					var assignment = binding as MemberAssignment;
+					if (assignment == null)
+					{
+						throw new ArgumentException("Only assignments bindings are supported");
+					}
+					var constant = assignment.Expression as ConstantExpression;
+					if (constant == null)
+					{
+						throw new ArgumentException("Only constant expressions are supported");
+					}
+
+					object value = constant.Value;
+					var property = assignment.Member as PropertyInfo;
+					if (property != null)
+					{
+						namedProperties.Add(property);
+						propertyValues.Add(value);
+					}
+					else
+					{
+						var field = assignment.Member as FieldInfo;
+						if (field != null)
+						{
+							namedFields.Add(field);
+							fieldValues.Add(value);
+						}
+						else
+						{
+							throw new ArgumentException("Only property and field assignments are supported");
+						}
+					}
+				}
+			}
+
+			var ctorArguments = new List<object>();
+			foreach (var arg in newExpression.Arguments)
+			{
+				var constant = arg as ConstantExpression;
+				if (constant == null)
+				{
+					throw new ArgumentException("Only constant expressions are supported");
+				}
+				ctorArguments.Add(constant.Value);
+			}
+
+			return new CustomAttributeInfo(
+				newExpression.Constructor,
+				ctorArguments.ToArray(),
+				namedProperties.ToArray(),
+				propertyValues.ToArray(),
+				namedFields.ToArray(),
+				fieldValues.ToArray());
+		}
+
+		internal CustomAttributeBuilder Builder
+		{
+			get { return builder; }
+		}
+
+		public bool Equals(CustomAttributeInfo other)
+		{
+			if (ReferenceEquals(null, other)) return false;
+			if (ReferenceEquals(this, other)) return true;
+			return con.Equals(other.con) &&
+				constructorArgs.SequenceEqual(other.constructorArgs) &&
+				AreMembersEquivalent(properties, other.properties) &&
+				AreMembersEquivalent(fields, other.fields);
+		}
+
+		public override bool Equals(object obj)
+		{
+			if (ReferenceEquals(null, obj)) return false;
+			if (ReferenceEquals(this, obj)) return true;
+			if (obj.GetType() != this.GetType()) return false;
+			return Equals((CustomAttributeInfo)obj);
+		}
+
+		public override int GetHashCode()
+		{
+			unchecked
+			{
+				int hashCode = con.GetHashCode();
+				hashCode = (hashCode*397) ^ CombineConstructorArgHashCodes(constructorArgs);
+				hashCode = (hashCode*397) ^ CombineMemberHashCodes(properties);
+				hashCode = (hashCode*397) ^ CombineMemberHashCodes(fields);
+				return hashCode;
+			}
+		}
+
+		private static bool AreMembersEquivalent(IDictionary<string, object> x, IDictionary<string, object> y)
+		{
+			foreach (var kvp in x)
+			{
+				object value;
+				if (!y.TryGetValue(kvp.Key, out value))
+					return false;
+				if (!Equals(kvp.Value, value))
+					return false;
+			}
+			return true;
+		}
```

Comment:
I think most of the comparisons can be short-circuited by checking the length of all the collections before even checking any contents, that would fix this problem.